README.md
# ui-items

Copyright (C) 2017-2018 The Open Library Foundation

This software is distributed under the terms of the Apache License,
Version 2.0. See the file "[LICENSE](LICENSE)" for more information.

## Introduction

This is a [Stripes](https://github.com/folio-org/stripes-core/) UI module for administrating bibliographic items.

## Additional information

See the related [ui-users](https://github.com/folio-org/ui-users) module.

Other [modules](http://dev.folio.org/source-code/#client-side).

See project [UIIT](https://issues.folio.org/browse/UIIT)
at the [FOLIO issue tracker](http://dev.folio.org/community/guide-issues).

Other FOLIO Developer documentation is at [dev.folio.org](http://dev.folio.org/)


CHANGELOG.md
# Change history for ui-items

## DEPRECATED. See ui-inventory.

## 1.12.0 (IN PROGRESS)

* Dependency on inventory: 5.0
* Support permission for loan-type maintenance. Fixes UIIT-38.
* Rely on new-style permission name to guard material-types settings. Fixes UIIT-39.
* Use `<ControlledVocab>` from stripes-smart-components instead of `<AuthorityList>` from stripes-components. See STSMACOM-6. Requires stripes-smart-components v1.0.1. Fixes UIIT-74, UIIT-76.
* Use PropTypes, not React.PropTypes. Refs FOLIO-427.
* Update stripes-components, stripes-connect versions. Refs STRIPES-501. 

## [1.11.0](https://github.com/folio-org/ui-items/tree/v1.11.0) (2017-09-01)
[Full Changelog](https://github.com/folio-org/ui-items/compare/v1.10.1...v1.11.0)

* Add permissions `settings.material-types.all`, `settings.loan-types.all` and `settings.items.enabled`. Fixes UIIT-59.
* Upgrade dependencies to stripes-components 1.7.0, stripes-connect 2.7.0 and stripes-core 2.7.0.

## [1.10.1](https://github.com/folio-org/ui-items/tree/v1.10.1) (2017-08-31)
[Full Changelog](https://github.com/folio-org/ui-items/compare/v1.10.0...v1.10.1)

* Update permissions. Completes STRIPES-435 for ui-items.

## [1.10.0](https://github.com/folio-org/ui-items/tree/v1.10.0) (2017-08-30)
[Full Changelog](https://github.com/folio-org/ui-items/compare/v1.9.0...v1.10.0)

* Make settings.material-types.all and settings.loan-types.all visible. For UIIT-38 and UIIT-39.
* Add test suites, carried over from ui-testing. FOLIO-800.

## [1.9.0](https://github.com/folio-org/ui-items/tree/v1.9.0) (2017-08-23)
[Full Changelog](https://github.com/folio-org/ui-items/compare/v1.8.0...v1.9.0)

* Do not mention spelling in "No Results Found" message if there is no search-term. Fixes UIIT-54.
* Switch from props.data to props.resources. Fixes UIIT-58.
* Assign IDs to more UI elements for automated testing. STRIPES-300.

## [1.8.0](https://github.com/folio-org/ui-items/tree/v1.8.0) (2017-08-03)
[Full Changelog](https://github.com/folio-org/ui-items/compare/v1.7.0...v1.8.0)

* Added Right Click to open item in new tab. Fixes UIIT-55.
* Move "New Item" button from filter pane to results pane. Fixes UIIT-37.
* New permission `settings.material-types.all`. Fixes UIIT-39.
* New permission `settings.loan-types.all`. Fixes UIIT-38.
* Add [`<SRStatus>`](https://github.com/folio-org/stripes-components/tree/master/lib/SRStatus) component to main module page. After search is performed, screen readers will announce the number of results. Integrates STCOM-3.
* Add "Skip to Results" [`<FocusLink>`](https://github.com/folio-org/stripes-components/tree/master/lib/FocusLink) component to search input (accessibility feature - tab twice on input to see it appear after 'clear search' button. When it is focused, pressing the 'enter' key will focus the results list). Integrates STCOM-7.
* Change title in results pane. See comments in UIU-125.
* Do not mention search-term in "No Results Found" message if there is none. Fixes UIIT-51.
* Avoid "no serverChoiceIndexes defined" error when running against recent mod-inventory-storage. Fixes UIIT-50.
* Material-type sorting is now by MT name, not ID. Fixes UIIT-33.
* The clear-search button goes back to Users starting state. Fixes UIIT-52.
* Explicitly reset component state in onClearSearch. Fixes UIIT-53.
* Update stripes-components to 1.4.0.

## [1.7.0](https://github.com/folio-org/ui-items/tree/v1.7.0) (2017-07-11)
[Full Changelog](https://github.com/folio-org/ui-items/compare/v1.6.0...v1.7.0)

* Upgrade stripes-components dependency to 1.3.0, needed for import of AuthorityList
* Sort-of-reinstate sorting by material-type. For deep reasons (see FOLIO-677) this can't work properly, but at least now each set of same-MT items are grouped together, even if the order of the groupings is apparently random. Fixes UIIT-28.
* When a new record is created, navigate to it. Fixes UIIT-29.
* Validation of item records. UIIT-9
* Add okapiInterfaces and permissionSets in `package.json`. Fixes UIIT-35.
* Asynchronous uniqueness-check for barcode. Fixes UIIT-32.
* Prompt on navigation away from dirty Item form. Refs UIIT-36.
* Eliminate uses of `componentWillMount` to provide initial values to local resources. Part of STRIPES-433.

## [1.6.0](https://github.com/folio-org/ui-items/tree/v1.6.0) (2017-06-14)
[Full Changelog](https://github.com/folio-org/ui-items/compare/v1.5.0...v1.6.0)

* Port from old `makePathFunction` to `makeQueryFunction`.
* Support reverse sorting. Fixes UIIT-25.
* Stable sorting: when moving from one sort criterion to another, the old one is retained as the secondary key. Fixes UIIT-26.
* Debounce search requests by 250 ms. Fixes UIIT-27.
* Upgrade stripes-components dependency to v0.15.0, needed for stable sorting.

## [1.5.0](https://github.com/folio-org/ui-items/tree/v1.5.0) (2017-06-08)
[Full Changelog](https://github.com/folio-org/ui-items/compare/v1.4.0...v1.5.0)

* Updated dependency on stripes-components to 0.12.0
* Added appropriate props for infinite-scroll functionality of results list. (STRIPES-361)
* The editMode boolean is now in a stripes-core local resource instead of a piece of React component state. This makes edit-mode persistent because it's in the Redux store -- so navigating away from an edit-item page to another app, then returning to the Items app, will remain in edit mode. Fixes UIIT-22.
* Use `props.resources` (from stripes-connect v2.2.0) to determine actual hit count. Fixes UIIT-21.
* Require stripes-components v0.11.2, stripes-core v1.9.0 and stripes-connect v2.2.0.

## [1.4.0](https://github.com/folio-org/ui-items/tree/v1.4.0) (2017-06-06)
[Full Changelog](https://github.com/folio-org/ui-items/compare/v1.3.0...v1.4.0)

* Loan & material types are now displayed/saved correctly when viewing/editing an item. Fixes STRIPES-355.
* Settings panes are sorted alphabetically. Fixes ui-items bit of STRIPES-358.

Items.js
import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import Route from 'react-router-dom/Route';
import queryString from 'query-string';

import Pane from '@folio/stripes-components/lib/Pane';
import Paneset from '@folio/stripes-components/lib/Paneset';
import PaneMenu from '@folio/stripes-components/lib/PaneMenu';
import Button from '@folio/stripes-components/lib/Button';
import MultiColumnList from '@folio/stripes-components/lib/MultiColumnList';
import FilterPaneSearch from '@folio/stripes-components/lib/FilterPaneSearch';
import Layer from '@folio/stripes-components/lib/Layer';
import FilterGroups, { initialFilterState, onChangeFilter as commonChangeFilter } from '@folio/stripes-components/lib/FilterGroups';
import SRStatus from '@folio/stripes-components/lib/SRStatus';
import transitionToParams from '@folio/stripes-components/util/transitionToParams';
import makeQueryFunction from '@folio/stripes-components/util/makeQueryFunction';
import removeQueryParam from '@folio/stripes-components/util/removeQueryParam';
import craftLayerUrl from '@folio/stripes-components/util/craftLayerUrl';

import ItemForm from './ItemForm';
import ViewItem from './ViewItem';

import packageInfo from './package';

const INITIAL_RESULT_COUNT = 30;
const RESULT_COUNT_INCREMENT = 30;

const filterConfig = [
  {
    label: 'Material Types',
    name: 'item',
    cql: 'materialType.id',
    values: [], // will be filled in by componentWillUpdate
  },
];

class Items extends React.Component {
  static propTypes = {
    stripes: PropTypes.shape({
      connect: PropTypes.func.isRequired,
      logger: PropTypes.shape({
        log: PropTypes.func.isRequired,
      }).isRequired,
    }).isRequired,
    resources: PropTypes.shape({
      materialTypes: PropTypes.shape({
        records: PropTypes.arrayOf(PropTypes.object),
      }),
      loanTypes: PropTypes.shape({
        records: PropTypes.arrayOf(PropTypes.object),
      }),
      items: PropTypes.shape({
        hasLoaded: PropTypes.bool.isRequired,
        other: PropTypes.shape({
          totalRecords: PropTypes.number,
          total_records: PropTypes.number,
        }),
        isPending: PropTypes.bool.isPending,
        successfulMutations: PropTypes.arrayOf(
          PropTypes.shape({
            record: PropTypes.shape({
              id: PropTypes.string.isRequired,
            }).isRequired,
          }),
        ),
      }),
      itemCount: PropTypes.number,
    }).isRequired,
    history: PropTypes.shape({
      push: PropTypes.func.isRequired,
    }).isRequired,
    location: PropTypes.shape({
      pathname: PropTypes.string.isRequired,
      search: PropTypes.string,
    }).isRequired,
    match: PropTypes.shape({
      path: PropTypes.string.isRequired,
    }).isRequired,
    mutator: PropTypes.shape({
      items: PropTypes.shape({
        POST: PropTypes.func,
      }),
      itemCount: PropTypes.shape({
        replace: PropTypes.func,
      }),
    }).isRequired,
    okapi: PropTypes.object,
  };

  static manifest = Object.freeze({
    itemCount: { initialValue: INITIAL_RESULT_COUNT },
    items: {
      type: 'okapi',
      records: 'items',
      path: 'inventory/items',
      recordsRequired: '%{itemCount}',
      perRequest: RESULT_COUNT_INCREMENT,
      GET: {
        params: {
          query: makeQueryFunction(
            'cql.allRecords=1',
            'materialType.name="$QUERY" or barcode="$QUERY*" or title="$QUERY*"',
            { 'Material Type': 'materialType.name' },
            filterConfig,
          ),
        },
        staticFallback: { params: {} },
      },
    },
    materialTypes: {
      type: 'okapi',
      path: 'material-types',
      records: 'mtypes',
    },
    loanTypes: {
      type: 'okapi',
      path: 'loan-types',
      records: 'loantypes',
    },
  });

  constructor(props) {
    super(props);

    const query = props.location.search ? queryString.parse(props.location.search) : {};
    this.state = {
      filters: initialFilterState(filterConfig, query.filters),
      selectedItem: {},
      searchTerm: query.query || '',
      sortOrder: query.sort || '',
    };

    this.onClearSearch = this.onClearSearch.bind(this);
    this.onSort = this.onSort.bind(this);
    this.onSelectRow = this.onSelectRow.bind(this);
    this.onClickAddNewItem = this.onClickAddNewItem.bind(this);
    this.onClickCloseNewItem = this.onClickCloseNewItem.bind(this);
    this.onChangeSearch = this.onChangeSearch.bind(this);

    this.onChangeFilter = commonChangeFilter.bind(this);
    this.transitionToParams = transitionToParams.bind(this);
    this.removeQueryParam = removeQueryParam.bind(this);
    this.craftLayerUrl = craftLayerUrl.bind(this);

    this.collapseDetails = this.collapseDetails.bind(this);
    this.connectedViewItem = props.stripes.connect(ViewItem);
    const logger = props.stripes.logger;
    this.log = logger.log.bind(logger);

    this.anchoredRowFormatter = this.anchoredRowFormatter.bind(this);

    this.resultsList = null;
    this.SRStatus = null;
  }

  componentWillReceiveProps(nextProps) {
    const resource = this.props.resources.items;
    if (resource) {
      const sm = nextProps.resources.items.successfulMutations;
      if (sm.length > resource.successfulMutations.length) {
        this.onSelectRow(undefined, { id: sm[0].record.id });
      }
    }

    if (resource && resource.isPending && !nextProps.resources.items.isPending) {
      this.log('event', 'new search-result');
      const resultAmount = nextProps.resources.items.other.totalRecords;
      this.SRStatus.sendMessage(`Search returned ${resultAmount} result${resultAmount !== 1 ? 's' : ''}`);
    }
  }

  componentWillUpdate() {
    const mt = (this.props.resources.materialTypes || {}).records || [];
    if (mt && mt.length) {
      filterConfig[0].values = mt.map(rec => ({ name: rec.name, cql: rec.id }));
    }
  }

  onClearSearch() {
    const path = (_.get(packageInfo, ['stripes', 'home']) ||
                  _.get(packageInfo, ['stripes', 'route']));
    this.setState({
      searchTerm: '',
      sortOrder: 'title',
      filters: {},
    });
    this.log('action', `cleared search: navigating to ${path}`);
    this.props.history.push(path);
  }

  onSort(e, meta) {
    const newOrder = meta.alias;
    const oldOrder = this.state.sortOrder || '';

    const orders = oldOrder ? oldOrder.split(',') : [];
    if (orders[0] && newOrder === orders[0].replace(/^-/, '')) {
      orders[0] = `-${orders[0]}`.replace(/^--/, '');
    } else {
      orders.unshift(newOrder);
    }

    const sortOrder = orders.slice(0, 2).join(',');
    this.log('action', `sorted by ${sortOrder}`);
    this.setState({ sortOrder });
    this.transitionToParams({ sort: sortOrder });
  }

  onSelectRow(e, meta) {
    const itemId = meta.id;
    this.log('action', `clicked item ${itemId}`);
    this.setState({ selectedItem: meta });
    this.props.history.push(`/items/view/${itemId}${this.props.location.search}`);
  }

  onClickAddNewItem(e) {
    if (e) e.preventDefault();
    this.log('action', 'clicked "add new item"');
    this.transitionToParams({ layer: 'create' });
  }

  onClickCloseNewItem(e) {
    if (e) e.preventDefault();
    this.log('action', 'clicked "close new item"');
    this.removeQueryParam('layer');
  }

  onChangeFilter = (e) => {
    this.props.mutator.itemCount.replace(INITIAL_RESULT_COUNT);
    this.commonChangeFilter(e);
  }

  onChangeSearch(e) {
    this.props.mutator.itemCount.replace(INITIAL_RESULT_COUNT);
    const query = e.target.value;
    this.setState({ searchTerm: query });
    this.log('action', `will search for '${query}'`);
    this.performSearch(query);
  }

  onNeedMore = () => {
    this.props.mutator.itemCount.replace(this.props.resources.itemCount + RESULT_COUNT_INCREMENT);
  }

  getRowURL(rowData) {
    return `/items/view/${rowData.id}${this.props.location.search}`;
  }

  performSearch = _.debounce((query) => {
    this.log('action', `searched for '${query}'`);
    this.transitionToParams({ query });
  }, 250);

  updateFilters(filters) { // provided for onChangeFilter
    this.transitionToParams({ filters: Object.keys(filters).filter(key => filters[key]).join(',') });
  }

  create(item) {
    // POST item record
    this.log('action', `Creating new item record: ${JSON.stringify(item)}`);
    this.props.mutator.items.POST(item);
    this.onClickCloseNewItem();
  }

  collapseDetails() {
    this.setState({
      selectedItem: {},
    });
    this.props.history.push(`${this.props.match.path}${this.props.location.search}`);
  }

  anchoredRowFormatter(
    { rowIndex,
      rowClass,
      rowData,
      cells,
      rowProps,
      labelStrings },
  ) {
    return (
      <a
        href={this.getRowURL(rowData)}
        key={`row-${rowIndex}`}
        aria-label={labelStrings && labelStrings.join('...')}
        role="listitem"
        className={rowClass}
        {...rowProps}
      >
        {cells}
      </a>
    );
  }

  render() {
    const { resources, stripes, location } = this.props;
    const query = location.search ? queryString.parse(location.search) : {};
    const items = (resources.items || {}).records || [];
    const materialTypes = (resources.materialTypes || {}).records || [];
    const loanTypes = (resources.loanTypes || {}).records || [];

    /* searchHeader is a 'custom pane header' */
    const searchHeader = <FilterPaneSearch id="SearchField" onChange={this.onChangeSearch} onClear={this.onClearSearch} resultsList={this.resultsList} value={this.state.searchTerm} />;
    const newItemButton = (
      <PaneMenu>
        <Button
          id="clickable-new-item"
          href={this.craftLayerUrl('create')}
          onClick={this.onClickAddNewItem}
          title="+ Item"
          buttonStyle="primary paneHeaderNewButton"
        >+ New
        </Button>
      </PaneMenu>
    );

    const resultsFormatter = {
      'Material Type': x => _.get(x, ['materialType', 'name'], ''),
      title: x => _.get(x, ['title'], ''),
      barcode: x => _.get(x, ['barcode'], ''),
      status: x => _.get(x, ['status', 'name']) || '--',
    };

    const maybeTerm = this.state.searchTerm ? ` for "${this.state.searchTerm}"` : '';
    const maybeSpelling = this.state.searchTerm ? 'spelling and ' : '';
    return (
      <Paneset>
        <SRStatus ref={(ref) => { this.SRStatus = ref; }} />
        {/* Filter Pane */}
        <Pane defaultWidth="16%" header={searchHeader}>
          <FilterGroups config={filterConfig} filters={this.state.filters} onChangeFilter={this.onChangeFilter} />
        </Pane>
        {/* Results Pane */}
        <Pane
          defaultWidth="fill"
          paneTitle={
            <div style={{ textAlign: 'center' }}>
              <strong>Items</strong>
              <div>
                <em>{this.props.resources.items && this.props.resources.items.hasLoaded ? this.props.resources.items.other.totalRecords : ''} Result{items.length === 1 ? '' : 's'} Found</em>
              </div>
            </div>
          }
          lastMenu={newItemButton}
        >
          <MultiColumnList
            contentData={items}
            selectedRow={this.state.selectedItem}
            rowMetadata={['title', 'id']}
            formatter={resultsFormatter}
            onRowClick={this.onSelectRow}
            onHeaderClick={this.onSort}
            onNeedMoreData={this.onNeedMore}
            visibleColumns={['Material Type', 'barcode', 'title', 'status']}
            sortOrder={this.state.sortOrder.replace(/^-/, '').replace(/,.*/, '')}
            sortDirection={this.state.sortOrder.startsWith('-') ? 'descending' : 'ascending'}
            isEmptyMessage={`No results found${maybeTerm}. Please check your ${maybeSpelling}filters.`}
            loading={this.props.resources.items ? this.props.resources.items.isPending : false}
            autosize
            virtualize
            ariaLabel="Item search results"
            rowFormatter={this.anchoredRowFormatter}
            containerRef={(ref) => { this.resultsList = ref; }}
          />
        </Pane>

        {/* Details Pane */}
        <Route
          path={`${this.props.match.path}/view/:itemid`}
          render={props => <this.connectedViewItem stripes={stripes} paneWidth="44%" onClose={this.collapseDetails} {...props} />}
        />
        <Layer isOpen={query.layer ? query.layer === 'create' : false} label="Add New Item Dialog">
          <ItemForm
            initialValues={{ available_material_types: materialTypes, available_loan_types: loanTypes, status: { name: 'Available' }, instanceId: 'dummy' }}
            onSubmit={(record) => { this.create(record); }}
            onCancel={this.onClickCloseNewItem}
            okapi={this.props.okapi}
          />
        </Layer>
      </Paneset>
    );
  }
}

export default Items;
